tests/basic.rs

```diff
@@ -99,57 +49,78 @@ pub async fn run_server(
-            match listener.accept().await {
-                Ok((stream, addr)) => {
-                    loop {
-                        match stream.readable().await {
-                            Ok(_) => {}
-                            Err(e) => { err_chan.send(e); return; }
-                        }
-                        let mut buf = [0; 8192];
-                        let n = match stream.try_read(&mut buf) {
-                            Ok(n) => n,
-                            Err(e) => { err_chan.send(e); break; }
-                        };
-                        match std::str::from_utf8(&buf[0..n]) {
-                            Ok(v) => {
-                                out_chan.send(v);
-                                loop {
-                                    match stream.writable().await {
-                                        Ok(_) => {}
-                                        Err(e) => { err_chan.send(e); return; }
-                                    }
-                                    match stream.try_write(TEST_MESSAGE.as_bytes()) {
-                                        Ok(n) => {
-                                            break;
-                                        }
-                                        Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
-                                            continue;
-                                        }
-                                        Err(e) => { err_chan.send(e); return; }
-                                    }
-                                }
-                                break;
-                            },
-                            Err(e) => {
-                                err_chan.send(
-                                    std::io::Error::new(
-                                        std::io::ErrorKind::Other,
-                                        "failed to parse utf8"
-                                    )
-                                );
-                                return;
-                            }
-                        }
-                    }
-                },
-                Err(e) => { err_chan.send(e); return; }
+    let listener = match dtls::listen("udp", "127.0.0.1", server_port, server_config).await {
+        Ok(listener) => listener,
+        Err(e) => {
+            err_chan.send(e);
+            return;
+        }
+    };
+    server_ready.send(());
+    let (stream, addr) = match listener.accept().await {
+        Ok((stream, addr)) => (stream, addr),
+        Err(e) => {
+            err_chan.send(e);
+            return;
+        }
+    };
+    // TODO make sure addr is the expected client
+    let mut buf = vec![0_u8; 8192];
+    loop {
+        match stream.readable().await {
+            Ok(_) => {}
+            Err(e) => {
+                err_chan.send(e);
+                return;
+            }
+        }
+        match stream.try_read(&mut buf) {
```

Comment:
Calling `from_utf8` on `&buf[0..n]` is resulting in "borrowed value does not live long enough". I am just still just learning about borrowing and my understanding is that the scope of `buf` is `68-124`, so I think I should be able to borrow that value here. What mistake am I making?

Reply:
I already restructured the code to avoid using channels before I saw this. I went back and was able to get it working with a `String` channel, though. Thank you for the suggestion!

My original intent with the channels was to keep the code structure and logic as similar as possible to the pion e2e tests. I might actually try to use channels again now that you've shown me the proper way to do it.

Reply:
Hack away😁